pkg/lvgl/contrib/lvgl.c

```diff
@@ -71,6 +71,10 @@ static void *_task_thread(void *arg)
         }
         else {
             if (IS_USED(MODULE_TOUCH_DEV))
+                /* When a touch device is in use, we cannot guarantee that the
+                   provides wake-up capabilities (stmpe811 cannot for instance).
+                   So in this case, the lvgl task handler thread is kept awake
+                   all the time. */
```

Comment:
That is odd - so the `stmpe811` requires active polling to get touch events?
The thing has an [interrupt pin](https://docs.rs-online.com/6daa/0900766b80edc3b7.pdf) and you even implemented that functionality.

It's just that in `tests/pkg_lvgl_touch/main.c` you don't configure an interrupt callback

```C
    stmpe811_init(&s_touch_dev, &stmpe811_params[0], NULL, NULL);
```

So how about

```C

static void _touch_event_cb(void *arg)
{
    (void)arg;
    lvgl_wakeup();
}

…

stmpe811_init(&s_touch_dev, &stmpe811_params[0], _touch_event_cb, NULL);
```

Reply:
That doesn't work (you can try).

Reply:
Looks like the interrupt polarity is set wrong in the driver.

This patch works for me:

```patch
diff --git a/drivers/stmpe811/stmpe811.c b/drivers/stmpe811/stmpe811.c
index a04a2b0810..43dd06a845 100644
--- a/drivers/stmpe811/stmpe811.c
+++ b/drivers/stmpe811/stmpe811.c
@@ -162,7 +162,7 @@ int stmpe811_init(stmpe811_t *dev, const stmpe811_params_t * params, touch_event
 
     if ((dev->params.int_pin != GPIO_UNDEF) && cb) {
         DEBUG("[stmpe811] init: configuring touchscreen interrupt\n");
-        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_RISING, cb, arg);
+        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_FALLING, cb, arg);
 
         /* Enable touchscreen interrupt */
         ret += i2c_write_reg(STMPE811_DEV_I2C, STMPE811_DEV_ADDR,
diff --git a/pkg/lvgl/contrib/lvgl.c b/pkg/lvgl/contrib/lvgl.c
index 795094340e..97c7d75e28 100644
--- a/pkg/lvgl/contrib/lvgl.c
+++ b/pkg/lvgl/contrib/lvgl.c
@@ -70,19 +70,11 @@ static void *_task_thread(void *arg)
             lv_task_handler();
         }
         else {
-            if (IS_USED(MODULE_TOUCH_DEV))
-                /* When a touch device is in use, we cannot guarantee that the
-                   provides wake-up capabilities (stmpe811 cannot for instance).
-                   So in this case, the lvgl task handler thread is kept awake
-                   all the time. */
-                lv_task_handler();
-            else {
-                /* Block after LVGL_ACTIVITY_PERIOD msec inactivity */
-                thread_flags_wait_one(LVGL_THREAD_FLAG);
-
-                /* trigger an activity so the task handler is called on the next loop */
-                lv_disp_trig_activity(NULL);
-            }
+            /* Block after LVGL_ACTIVITY_PERIOD msec inactivity */
+            thread_flags_wait_one(LVGL_THREAD_FLAG);
+
+            /* trigger an activity so the task handler is called on the next loop */
+            lv_disp_trig_activity(NULL);
         }
 
         xtimer_usleep(LVGL_TASK_HANDLER_DELAY_US);
diff --git a/tests/driver_stmpe811/main.c b/tests/driver_stmpe811/main.c
index e29361558a..1479e65e14 100644
--- a/tests/driver_stmpe811/main.c
+++ b/tests/driver_stmpe811/main.c
@@ -20,6 +20,7 @@
 
 #include <stdio.h>
 
+#include "mutex.h"
 #include "xtimer.h"
 
 #include "stmpe811.h"
@@ -27,18 +28,19 @@
 
 static void _touch_event_cb(void *arg)
 {
-    (void)arg;
     puts("Pressed!");
+    mutex_unlock(arg);
 }
 
 int main(void)
 {
+    mutex_t lock = MUTEX_INIT_LOCKED;
     stmpe811_t dev;
 
     puts("STMPE811 test application\n");
 
     printf("+------------Initializing------------+\n");
-    int ret = stmpe811_init(&dev, &stmpe811_params[0], _touch_event_cb, NULL);
+    int ret = stmpe811_init(&dev, &stmpe811_params[0], _touch_event_cb, &lock);
     if (ret != STMPE811_OK) {
         puts("[Error] Initialization failed");
         return 1;
@@ -51,6 +53,8 @@ int main(void)
     stmpe811_touch_state_t last_touch_state = current_touch_state;
 
     while (1) {
+        mutex_lock(&lock);
+
         stmpe811_read_touch_state(&dev, &current_touch_state);
         if (current_touch_state != last_touch_state) {
             if (current_touch_state == STMPE811_TOUCH_STATE_RELEASED) {
@@ -65,8 +69,6 @@ int main(void)
             stmpe811_read_touch_position(&dev, &position);
             printf("X: %i, Y:%i\n", position.x, position.y);
         }
-
-        xtimer_usleep(10 * US_PER_MS);
     }
 
     return 0;
diff --git a/tests/pkg_lvgl_touch/main.c b/tests/pkg_lvgl_touch/main.c
index 2c13ab05ac..db1ce99a3b 100644
--- a/tests/pkg_lvgl_touch/main.c
+++ b/tests/pkg_lvgl_touch/main.c
@@ -45,6 +45,14 @@ void btn_event_cb(lv_obj_t * btn, lv_event_t event)
     }
 }
 
+static void _touch_event_cb(void *arg)
+{
+    (void)arg;
+
+    puts("touch");
+    lvgl_wakeup();
+}
+
 int main(void)
 {
     /* Configure the generic display driver interface */
@@ -59,7 +67,7 @@ int main(void)
     s_screen.touch->driver = &stmpe811_touch_dev_driver;
 
     /* Initialize the concrete touch driver */
-    stmpe811_init(&s_touch_dev, &stmpe811_params[0], NULL, NULL);
+    stmpe811_init(&s_touch_dev, &stmpe811_params[0], _touch_event_cb, NULL);
 
     /* Initialize lvgl with the generic screen */
     lvgl_init(&s_screen);
```

Reply:
That was done initially like this but it doesn't work. I also tried to change the polarity but that doesn't make any difference.

You are missing one important point when testing: you have to wait +10s before trying to press again. Otherwise you are within the activity time frame and the lvgl thread is not blocked. But once it's blocked, no interrupt is raised by the driver and it cannot wakeup the thread.

Reply:
With the following diff, I have something that works:

<details>

```diff
diff --git a/drivers/stmpe811/stmpe811.c b/drivers/stmpe811/stmpe811.c
index a04a2b0810..8f2aa44dc4 100644
--- a/drivers/stmpe811/stmpe811.c
+++ b/drivers/stmpe811/stmpe811.c
@@ -162,7 +162,7 @@ int stmpe811_init(stmpe811_t *dev, const stmpe811_params_t * params, touch_event
 
     if ((dev->params.int_pin != GPIO_UNDEF) && cb) {
         DEBUG("[stmpe811] init: configuring touchscreen interrupt\n");
-        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_RISING, cb, arg);
+        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_FALLING, cb, arg);
 
         /* Enable touchscreen interrupt */
         ret += i2c_write_reg(STMPE811_DEV_I2C, STMPE811_DEV_ADDR,
@@ -179,6 +179,8 @@ int stmpe811_init(stmpe811_t *dev, const stmpe811_params_t * params, touch_event
         return -STMPE811_ERR_I2C;
     }
 
+    _reset_fifo(dev);
+
     /* Release I2C device */
     i2c_release(STMPE811_DEV_I2C);
 
@@ -262,8 +264,9 @@ int stmpe811_read_touch_state(const stmpe811_t *dev, stmpe811_touch_state_t *sta
         return -STMPE811_ERR_I2C;
     }
 
+    _clear_interrupt_status(dev);
     if ((val & STMPE811_TSC_CTRL_STA)) {
-        _clear_interrupt_status(dev);
+        // _clear_interrupt_status(dev);
         *state = STMPE811_TOUCH_STATE_PRESSED;
     }
     else {
diff --git a/pkg/lvgl/contrib/lvgl.c b/pkg/lvgl/contrib/lvgl.c
index 795094340e..02bc51e333 100644
--- a/pkg/lvgl/contrib/lvgl.c
+++ b/pkg/lvgl/contrib/lvgl.c
@@ -70,19 +70,19 @@ static void *_task_thread(void *arg)
             lv_task_handler();
         }
         else {
-            if (IS_USED(MODULE_TOUCH_DEV))
-                /* When a touch device is in use, we cannot guarantee that the
-                   provides wake-up capabilities (stmpe811 cannot for instance).
-                   So in this case, the lvgl task handler thread is kept awake
-                   all the time. */
-                lv_task_handler();
-            else {
+            // if (IS_USED(MODULE_TOUCH_DEV))
+            //     /* When a touch device is in use, we cannot guarantee that the
+            //        provides wake-up capabilities (stmpe811 cannot for instance).
+            //        So in this case, the lvgl task handler thread is kept awake
+            //        all the time. */
+            //     lv_task_handler();
+            // else {
                 /* Block after LVGL_ACTIVITY_PERIOD msec inactivity */
                 thread_flags_wait_one(LVGL_THREAD_FLAG);
 
                 /* trigger an activity so the task handler is called on the next loop */
                 lv_disp_trig_activity(NULL);
-            }
+            // }
         }
 
         xtimer_usleep(LVGL_TASK_HANDLER_DELAY_US);
diff --git a/tests/pkg_lvgl_touch/main.c b/tests/pkg_lvgl_touch/main.c
index 2c13ab05ac..71f4e094ee 100644
--- a/tests/pkg_lvgl_touch/main.c
+++ b/tests/pkg_lvgl_touch/main.c
@@ -37,6 +37,13 @@ static screen_dev_t s_screen;
 static ili9341_t s_disp_dev;
 static stmpe811_t s_touch_dev;
 
+static void _touch_event_cb(void *arg)
+{
+    (void)arg;
+    // puts("Pressed!");
+    lvgl_wakeup();
+}
+
 void btn_event_cb(lv_obj_t * btn, lv_event_t event)
 {
     (void)btn;
@@ -59,7 +66,7 @@ int main(void)
     s_screen.touch->driver = &stmpe811_touch_dev_driver;
 
     /* Initialize the concrete touch driver */
-    stmpe811_init(&s_touch_dev, &stmpe811_params[0], NULL, NULL);
+    stmpe811_init(&s_touch_dev, &stmpe811_params[0], _touch_event_cb, NULL);
 
     /* Initialize lvgl with the generic screen */
     lvgl_init(&s_screen);
```

</details>

- the lvgl thread wakes up and events are processed when the screen is pressed
- no invalid touch positions (even the first one), this can be checked using `tests/driver_stmpe811`